File: app/controllers/transcription_controller.rb
class TranscriptionController < ApplicationController
  def index
    @transcriptions = Transcription.order(created_at: :desc).limit(10)
  end

  def create
    Rails.logger.info("Creating new transcription with params: #{transcription_params.to_h.except(:audio_data)}")
    
    @transcription = Transcription.new(transcription_params)
    
    # Log audio data presence
    if transcription_params[:audio_data].present?
      Rails.logger.info("Audio data is present in the request")
      audio_data_length = transcription_params[:audio_data].length
      Rails.logger.info("Audio data length: #{audio_data_length} characters")
    else
      Rails.logger.warn("No audio data present in the request")
    end
    
    if @transcription.save
      Rails.logger.info("Transcription saved successfully with ID: #{@transcription.id}")
      Rails.logger.info("Transcription text content: #{@transcription.text_content}")
      
      # Check if audio file was attached
      if @transcription.audio_file.attached?
        Rails.logger.info("Audio file successfully attached to transcription")
        Rails.logger.info("Audio file content type: #{@transcription.audio_file.content_type}")
        Rails.logger.info("Audio file size: #{@transcription.audio_file.byte_size} bytes")
      else
        Rails.logger.warn("No audio file attached to transcription")
      end
      
      # Process the transcription with grammar analysis
      begin
        if Rails.env.development?
          # In development, perform the job immediately for faster feedback
          Rails.logger.info("Running grammar analysis job immediately (development mode)")
          GrammarAnalysisJob.perform_now(@transcription.id)
        else
          # In production, use background processing
          Rails.logger.info("Enqueueing grammar analysis job (production mode)")
          GrammarAnalysisJob.perform_later(@transcription.id)
        end
      rescue => e
        Rails.logger.error("Error processing grammar analysis job: #{e.message}")
        Rails.logger.error("Backtrace: #{e.backtrace.join('\n')}")
        # Continue with the response even if the job fails
      end
      
      respond_to do |format|
        format.turbo_stream
        format.html { redirect_to transcription_show_path(@transcription) }
      end
    else
      Rails.logger.error("Failed to save transcription: #{@transcription.errors.full_messages.join(', ')}")
      render :index, status: :unprocessable_entity
    end
  end

  def show
    @transcription = Transcription.find(params[:id])
  end
  
  # API endpoint for transcribing audio
  def transcribe
    begin
      Rails.logger.info("Received transcription request")
      
      # Get the audio data from the request
      audio_data = params[:audio_data]
      
      if audio_data.nil? || audio_data.empty?
        Rails.logger.error("No audio data received in request")
        render json: { error: "No audio data provided" }, status: :bad_request
        return
      end
      
      # Log audio data format
      if audio_data.is_a?(String) && audio_data.include?('base64')
        Rails.logger.info("Received base64 audio data, length: #{audio_data.length} characters")
        content_type = audio_data.match(/data:(.*);base64/)[1]
        Rails.logger.info("Audio content type: #{content_type}")
      elsif audio_data.respond_to?(:content_type)
        Rails.logger.info("Received file upload, content type: #{audio_data.content_type}")
      else
        Rails.logger.info("Received audio data of type: #{audio_data.class}")
      end
      
      Rails.logger.info("Starting transcription process")
      
      # Use the TranscriptionService to transcribe the audio
      transcription_service = TranscriptionService.new(audio_data)
      transcribed_text = transcription_service.transcribe
      
      Rails.logger.info("Transcription successful, text length: #{transcribed_text.length}")
      Rails.logger.info("Transcribed text: #{transcribed_text}")
      
      # Return the transcribed text as JSON
      render json: { text: transcribed_text }
    rescue => e
      Rails.logger.error("Error during transcription: #{e.message}")
      Rails.logger.error("Backtrace: #{e.backtrace.join('\n')}")
      render json: { error: "Transcription failed: #{e.message}" }, status: :unprocessable_entity
    end
  end

  private

  def transcription_params
    params.require(:transcription).permit(:audio_data, :text_content)
  end
end

File: app/javascript/controllers/transcription_controller.js
import { Controller } from "@hotwired/stimulus"

// Connects to data-controller="transcription"
export default class extends Controller {
  static targets = ["recordButton", "stopButton", "recordingStatus", "audioPreview", "audioPlayer", "transcriptionForm", "transcriptionText"]

  connect() {
    this.mediaRecorder = null
    this.audioChunks = []
    this.audioBlob = null
    this.audioUrl = null
  }

  startRecording() {
    // Request audio with specific constraints for better compatibility
    const constraints = {
      audio: {
        echoCancellation: true,
        noiseSuppression: true,
        autoGainControl: true
      }
    }
    
    navigator.mediaDevices.getUserMedia(constraints)
      .then(stream => {
        // Use WebM with Opus codec for better compatibility
        // If that's not supported, fall back to default
        const options = {
          mimeType: 'audio/webm;codecs=opus'
        }
        
        try {
          this.mediaRecorder = new MediaRecorder(stream, options)
          console.log('Using MIME type:', this.mediaRecorder.mimeType)
        } catch (e) {
          console.warn('WebM with Opus not supported, using default codec')
          this.mediaRecorder = new MediaRecorder(stream)
          console.log('Using MIME type:', this.mediaRecorder.mimeType)
        }
        
        this.audioChunks = []
        
        this.mediaRecorder.addEventListener('dataavailable', event => {
          if (event.data.size > 0) {
            this.audioChunks.push(event.data)
            console.log(`Received audio chunk: ${event.data.size} bytes`)
          }
        })
        
        this.mediaRecorder.addEventListener('stop', () => this.processRecording())
        
        // Request data every second to ensure we get data even if recording is interrupted
        this.mediaRecorder.start(1000)
        this.recordingStatusTarget.textContent = 'Recording...'
        this.recordButtonTarget.classList.add('hidden')
        this.stopButtonTarget.classList.remove('hidden')
      })
      .catch(err => {
        console.error('Error accessing microphone:', err)
        this.recordingStatusTarget.textContent = 'Error: Could not access microphone. Please ensure you have granted permission.'
      })
  }
  
  stopRecording() {
    if (this.mediaRecorder && this.mediaRecorder.state !== 'inactive') {
      this.mediaRecorder.stop()
      this.recordingStatusTarget.textContent = 'Processing audio...'
      this.stopButtonTarget.classList.add('hidden')
      this.recordButtonTarget.classList.remove('hidden')
      
      // Stop all audio tracks
      this.mediaRecorder.stream.getTracks().forEach(track => track.stop())
    }
  }
  
  processRecording() {
    // Determine the correct MIME type based on what was recorded
    const mimeType = this.mediaRecorder ? this.mediaRecorder.mimeType : 'audio/webm'
    console.log(`Creating blob with MIME type: ${mimeType}`)
    
    // Create a blob from the audio chunks with the correct MIME type
    this.audioBlob = new Blob(this.audioChunks, { type: mimeType })
    this.audioUrl = URL.createObjectURL(this.audioBlob)
    
    // Log blob size and type
    console.log(`Audio blob size: ${this.audioBlob.size} bytes`)
    console.log(`Audio blob type: ${this.audioBlob.type}`)
    
    if (this.audioBlob.size === 0) {
      console.error('Audio blob is empty!')
      this.recordingStatusTarget.textContent = 'Error: No audio data recorded. Please try again.'
      return
    }
    
    // Set the audio player source
    this.audioPlayerTarget.src = this.audioUrl
    this.audioPreviewTarget.classList.remove('hidden')
    
    // Add an event listener to check if the audio can be played
    this.audioPlayerTarget.addEventListener('canplay', () => {
      console.log('Audio can be played in the preview player')
    })
    
    this.audioPlayerTarget.addEventListener('error', (e) => {
      console.error('Error loading audio in preview player:', e)
    })
    
    // Convert blob to base64 for form submission
    const reader = new FileReader()
    reader.readAsDataURL(this.audioBlob)
    reader.onloadend = () => {
      const base64data = reader.result
      console.log(`Base64 data length: ${base64data.length} characters`)
      console.log(`Base64 data starts with: ${base64data.substring(0, 50)}...`)
      
      // Store the audio data in the form
      const audioDataInput = document.getElementById('transcription_audio_data')
      audioDataInput.value = base64data
      
      // Store the audio data as a data attribute on the form for later retrieval
      this.transcriptionFormTarget.dataset.audioData = base64data
      
      // Also store the MIME type for reference
      this.transcriptionFormTarget.dataset.audioMimeType = mimeType
      
      // Verify the data was set
      console.log(`Audio data input value length: ${audioDataInput.value.length} characters`)
      
      // Now transcribe the audio
      this.transcribeAudio()
    }
  }
  
  async transcribeAudio() {
    // Send the audio to the server for transcription using OpenAI Whisper API
    this.recordingStatusTarget.textContent = 'Transcribing audio...'
    
    try {
      // Get the audio data from the form
      const audioData = document.getElementById('transcription_audio_data').value
      console.log(`Sending audio data for transcription, length: ${audioData.length} characters`)
      
      // Create a FormData object to send to the server
      const formData = new FormData()
      formData.append('audio_data', audioData)
      
      // Log that we're about to send the request
      console.log('Sending transcription request to server...')
      
      // Send the audio data to the server for transcription
      const response = await fetch('/transcription/transcribe', {
        method: 'POST',
        headers: {
          'X-CSRF-Token': document.querySelector('meta[name="csrf-token"]').content,
          'Accept': 'application/json'
        },
        body: formData
      })
      
      console.log(`Received response with status: ${response.status}`)
      
      if (!response.ok) {
        const errorText = await response.text()
        console.error('Error response:', errorText)
        throw new Error(`Transcription failed with status ${response.status}: ${errorText}`)
      }
      
      const result = await response.json()
      console.log('Transcription result:', result)
      
      if (result.error) {
        throw new Error(result.error)
      }
      
      // Update the form and display with the transcribed text
      const textContentInput = document.getElementById('transcription_text_content')
      textContentInput.value = result.text
      this.transcriptionTextTarget.textContent = result.text
      this.transcriptionFormTarget.classList.remove('hidden')
      
      console.log('Form updated with transcribed text')
      console.log('Text content input value:', textContentInput.value)
      
      // Add an event listener to the form to ensure audio data is included when the form is submitted
      this.transcriptionFormTarget.addEventListener('submit', this.ensureAudioDataIncluded.bind(this), { once: true })
      
      this.recordingStatusTarget.textContent = 'Transcription complete. You can now save and analyze the grammar.'
    } catch (error) {
      console.error('Error during transcription:', error)
      this.recordingStatusTarget.textContent = `Error: ${error.message}. Please try again.`
    }
  }

  // Ensure audio data is included in the form submission
  ensureAudioDataIncluded(event) {
    // Get the stored audio data
    const audioData = this.transcriptionFormTarget.dataset.audioData
    
    if (audioData) {
      console.log('Adding audio data to form submission')
      const audioDataInput = document.getElementById('transcription_audio_data')
      
      // Always restore from saved data to ensure we have the complete audio data
      console.log('Ensuring audio data is included in form submission')
      audioDataInput.value = audioData
      
      // Log the data being submitted
      console.log(`Form submission audio data length: ${audioDataInput.value.length} characters`)
      console.log(`Audio MIME type: ${this.transcriptionFormTarget.dataset.audioMimeType || 'unknown'}`)
      
      // Add a hidden field with the MIME type if it exists
      if (this.transcriptionFormTarget.dataset.audioMimeType) {
        let mimeTypeInput = document.getElementById('audio_mime_type')
        if (!mimeTypeInput) {
          mimeTypeInput = document.createElement('input')
          mimeTypeInput.type = 'hidden'
          mimeTypeInput.id = 'audio_mime_type'
          mimeTypeInput.name = 'audio_mime_type'
          this.transcriptionFormTarget.appendChild(mimeTypeInput)
        }
        mimeTypeInput.value = this.transcriptionFormTarget.dataset.audioMimeType
      }
    } else {
      console.warn('No audio data found for form submission')
    }
  }
}
